lightning/src/blinded_path/payment.rs

```diff
@@ -0,0 +1,350 @@
+		/// Constraints for the receiver of this payment.
+		payment_constraints: PaymentConstraints,
+	},
+}
+
+impl BlindedPaymentTlvs {
+	// The fee used to get from the current hop to the next hop in the path.
+	fn fee_base_msat(&self) -> u32 {
+		match self {
+			Self::Forward { payment_relay, .. } => payment_relay.fee_base_msat,
+			_ => 0,
+		}
+	}
+	// The fee used to get from the current hop to the next hop in the path.
+	fn fee_proportional_millionths(&self) -> u32 {
+		match self {
+			Self::Forward { payment_relay, .. } => payment_relay.fee_proportional_millionths,
+			_ => 0,
+		}
+	}
+	// The delta used to get from the current hop to the next hop in the path.
+	fn cltv_expiry_delta(&self) -> u16 {
+		match self {
+			Self::Forward { payment_relay, .. } => payment_relay.cltv_expiry_delta,
+			_ => 0,
+		}
+	}
+	fn htlc_minimum_msat(&self) -> u64 {
+		match self {
+			Self::Forward { payment_constraints, .. } | Self::Receive { payment_constraints, .. } =>
+				payment_constraints.htlc_minimum_msat,
+		}
+	}
+	fn features(&self) -> Option<&BlindedHopFeatures> {
+		match self {
+			Self::Forward { features, .. } => Some(&features),
+			Self::Receive { .. } => None,
+		}
+	}
+}
+
+/// Parameters for relaying over a given [`BlindedHop`].
+///
+/// [`BlindedHop`]: crate::blinded_path::BlindedHop
+pub struct PaymentRelay {
+	/// Number of blocks subtracted from an incoming HTLC's `cltv_expiry` for this [`BlindedHop`].
+	///
+	///[`BlindedHop`]: crate::blinded_path::BlindedHop
+	pub cltv_expiry_delta: u16,
+	/// Liquidity fee charged (in millionths of the amount transferred) for relaying a payment over
+	/// this [`BlindedHop`], (i.e., 10,000 is 1%).
+	///
+	///[`BlindedHop`]: crate::blinded_path::BlindedHop
+	pub fee_proportional_millionths: u32,
+	/// Base fee charged (in millisatoshi) for relaying a payment over this [`BlindedHop`].
+	///
+	///[`BlindedHop`]: crate::blinded_path::BlindedHop
+	pub fee_base_msat: u32,
+}
+
+/// Constraints for relaying over a given [`BlindedHop`].
+///
+/// [`BlindedHop`]: crate::blinded_path::BlindedHop
+pub struct PaymentConstraints {
+	/// The maximum total CLTV delta that is acceptable when relaying a payment over this
+	/// [`BlindedHop`].
+	///
+	///[`BlindedHop`]: crate::blinded_path::BlindedHop
+	pub max_cltv_expiry: u32,
+	/// The minimum value, in msat, that may be relayed over this [`BlindedHop`].
+	pub htlc_minimum_msat: u64,
+}
+
+impl Writeable for BlindedPaymentTlvs {
+	fn write<W: Writer>(&self, w: &mut W) -> Result<(), io::Error> {
+		// TODO: write padding
+		match self {
+			Self::Forward { short_channel_id, payment_relay, payment_constraints, features } => {
+				encode_tlv_stream!(w, {
+					(2, short_channel_id, required),
+					(10, payment_relay, required),
+					(12, payment_constraints, required),
+					(14, features, required),
+				});
+			},
+			Self::Receive { payment_secret, payment_constraints } => {
+				encode_tlv_stream!(w, {
+					(6, payment_secret, required),
```

Comment:
Hmm, I'm not sure if it's necessary to use custom range because the `payment_secret` is valid as a `path_id` (TLV type 6). If we put it in the custom range, would we need a bLIP to reserve the TLV type?